File: leetcode/dp/63. Unique Paths II/js/main.js
// Time O(M * N)
// Space O(M * N)
const uniquePathsWithObstacles = grid => {
  if (grid[0][0] === 1) {
    return 0;
  }

  let n = grid.length;
  let m = grid[0].length;

  let matrix = Array(n)
    .fill(null)
    .map(() => Array(m).fill(0));

  matrix[0][0] = 1;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (i === 0 && j === 0) {
        continue;
      }

      if (grid[i][j] === 1) {
        matrix[i][j] = 0;
      } else if (i === 0) {
        matrix[i][j] = matrix[i][j - 1];
      } else if (j === 0) {
        matrix[i][j] = matrix[i - 1][j];
      } else {
        matrix[i][j] = matrix[i - 1][j] + matrix[i][j - 1];
      }
    }
  }

  return matrix[n - 1][m - 1];
};

File: leetcode/dp/63. Unique Paths II/py/main.py
# Time O(N^2)
# Space O(N^2)
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid):
        n, m = len(obstacleGrid), len(obstacleGrid[0])

        dp = [[0] * (m + 1) for _ in range(n + 1)]
                    
        for i in range(1, n + 1):
          for j in range(1, m + 1):
            if obstacleGrid[i - 1][j - 1] == 1:
              continue
            if i == 1 and j == 1:
                dp[i][j] = 1
            else:
                dp[i][j] = dp[i-1][j] + dp[i][j - 1]
                
        return dp[n][m]
            